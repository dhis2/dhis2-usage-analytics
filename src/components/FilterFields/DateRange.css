.uaa-date-field {
    position: relative;
}

.uaa-date-field.with-error {
    margin-bottom: 0;
    position: relative;
}

.uaa-date-field input {
    display: block;
    box-sizing: border-box;
    font-size: 16px;
    font-family: Roboto, sans-serif;
    width: 100%;
    height: 56px;
    border: none;
    padding: 15px 14px 0 8px;
    border-radius: 5px;
    background-color: rgba(0, 0, 10, 0.05);
}

.uaa-date-field input:focus,
.uaa-date-field input:hover {
    border-bottom: 2px solid #00796b;
    padding-top: 17px;
}

.uaa-date-field input:focus {
    outline: none;
}

.uaa-date-field label {
    position: absolute;
    top: 8px;
    left: 14px;
    font-size: 12px;
}
.uaa-date-input-error {
    color: #e53935;
    font-size: 12px;
    display: block;
    /* Will be rendered in the margin of the date-field
       and not increase the date-field's height */
    height: 0;
}
