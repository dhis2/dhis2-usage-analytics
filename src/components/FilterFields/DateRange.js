import React, { Component } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import debounce from 'lodash.debounce'
import './DateRange.css'

export const START_DATE = 'startDate'
export const END_DATE = 'endDate'
export const ERROR_PATTERN = i18n.t('Please use the format yyyy-mm-dd')
export const ERROR_MISSING_DATE = i18n.t('Please fill in both dates')
export const ERROR_END_BEFORE_START = i18n.t('End date is before start date')
export const ERROR_START_AFTER_END = i18n.t('Start date is after end date')

// Very basic date input. If browser doesn't support input type date,
// it will use the pattern validation to display an error
class DateRange extends Component {
    constructor(props) {
        super(props)
        this.state = {
            startDateError: null,
            endDateError: null,
        }
        this.onStartDateChange = this.onStartDateChange.bind(this)
        this.onEndDateChange = this.onEndDateChange.bind(this)
        this.updateUsageData = debounce(props.updateUsageData, 250)
    }

    onStartDateChange(event) {
        this.onChange(START_DATE, event.target.value)
    }
    onEndDateChange(event) {
        this.onChange(END_DATE, event.target.value)
    }

    onChange(key, value) {
        const { updateFilter, updateUsageData, ...dateRange } = this.props
        const errorKey = `${key}Error`
        const otherErrorKey =
            key === START_DATE ? `${END_DATE}Error` : `${START_DATE}Error`
        const error = this.getError({ ...dateRange, [key]: value }, key)
        const otherError =
            this.state[otherErrorKey] === ERROR_PATTERN ? ERROR_PATTERN : null

        if (this.state[errorKey] !== error) {
            this.setState({
                [errorKey]: error,
                [otherErrorKey]: otherError,
            })
        }

        updateFilter(key, value)

        if (!error) {
            this.updateUsageData()
        }
    }

    getError(dateRange, key) {
        if (!/[0-9]{4}-[0-9]{2}-[0-9]{2}/.test(dateRange[key])) {
            return ERROR_PATTERN
        }

        if (!dateRange.startDate || !dateRange.endDate) {
            return ERROR_MISSING_DATE
        }

        if (dateRange.endDate < dateRange.startDate) {
            return key === START_DATE
                ? ERROR_START_AFTER_END
                : ERROR_END_BEFORE_START
        }

        return null
    }

    render() {
        const { startDate, endDate } = this.props
        const { startDateError, endDateError } = this.state

        return (
            <div className="uaa-date-range">
                <div className="uaa-date-field uaa-fieldwrap">
                    <label>{i18n.t('Start Date')}</label>
                    <input
                        className={START_DATE}
                        type="date"
                        value={startDate}
                        onChange={this.onStartDateChange}
                    />
                    {startDateError && (
                        <span className={`uaa-date-input-error ${START_DATE}`}>
                            {startDateError}
                        </span>
                    )}
                </div>
                <div className="uaa-date-field uaa-fieldwrap">
                    <label>{i18n.t('End date')}</label>
                    <input
                        className={END_DATE}
                        type="date"
                        value={endDate}
                        onChange={this.onEndDateChange}
                    />
                    {endDateError && (
                        <span className={`uaa-date-input-error ${END_DATE}`}>
                            {endDateError}
                        </span>
                    )}
                </div>
            </div>
        )
    }
}

DateRange.propTypes = {
    startDate: PropTypes.string,
    endDate: PropTypes.string,
    updateFilter: PropTypes.func.isRequired,
    updateUsageData: PropTypes.func.isRequired,
}

export default DateRange
