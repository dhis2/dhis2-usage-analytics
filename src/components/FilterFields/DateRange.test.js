import React from 'react'
import { shallow } from 'enzyme'
import DateRange, {
    START_DATE,
    END_DATE,
    ERROR_PATTERN,
    ERROR_MISSING_DATE,
    ERROR_START_AFTER_END,
    ERROR_END_BEFORE_START,
} from './DateRange'

jest.mock('lodash.debounce', () => fn => fn)

describe('<DateRange/>', () => {
    const updateFilter = jest.fn()
    const updateUsageData = jest.fn()

    const defaultProps = {
        startDate: '2018-09-30',
        endDate: '2018-12-30',
        updateFilter,
        updateUsageData,
    }
    const NOT_A_DATE = 'This is not a date'

    const wrapper = shallow(<DateRange {...defaultProps} />)
    const startDateInput = wrapper.find(`.${START_DATE}`)
    const endDateInput = wrapper.find(`.${END_DATE}`)

    afterEach(() => {
        jest.clearAllMocks()
    })

    it('matches the snapshot with default props', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('renders a date input for startDate', () => {
        expect(startDateInput.prop('type')).toEqual('date')
    })
    it('renders a date input for endDate', () => {
        expect(endDateInput.prop('type')).toEqual('date')
    })
    it('calls updateFilterAndGetData when startDate changes into a valid value', () => {
        const value = '2018-09-20'
        startDateInput.simulate('change', { target: { value } })
        expect(updateUsageData).toHaveBeenCalledTimes(1)
    })
    it('calls updateFilterAndGetData when endDate changes into a valid value', () => {
        const value = '2019-01-20'
        endDateInput.simulate('change', { target: { value } })
        expect(updateUsageData).toHaveBeenCalledTimes(1)
    })
    it('calls updateFilter when startDate changes', () => {
        const value = '2019-12-31'
        startDateInput.simulate('change', { target: { value } })
        expect(updateFilter).toBeCalledWith(START_DATE, value)
    })
    it('calls updateFilter when endDate changes', () => {
        const value = '2016-01-20'
        endDateInput.simulate('change', { target: { value } })
        expect(updateFilter).toBeCalledWith(END_DATE, value)
    })
    it('renders a startDate pattern error when startDate changes into an invalid value', () => {
        const value = NOT_A_DATE
        startDateInput.simulate('change', { target: { value } })
        expect(
            wrapper.find(`span.uaa-date-input-error.${START_DATE}`)
        ).toHaveText(ERROR_PATTERN)
    })
    it('renders a missing date error when startDate changes into a valid value but endDate is empty', () => {
        const defaultProps = {
            updateFilter,
            updateUsageData,
        }
        const wrapper = shallow(<DateRange {...defaultProps} />)
        const startDateInput = wrapper.find(`.${START_DATE}`)

        startDateInput.simulate('change', { target: { value: '2019-12-30' } })

        expect(
            wrapper.find(`span.uaa-date-input-error.${START_DATE}`)
        ).toHaveText(ERROR_MISSING_DATE)
    })
    it('renders a start-after-end error when start date become greater than end date', () => {
        const value = '2020-01-01'
        startDateInput.simulate('change', { target: { value } })
        expect(
            wrapper.find(`span.uaa-date-input-error.${START_DATE}`)
        ).toHaveText(ERROR_START_AFTER_END)
    })
    it('renders a end-before-start error when end date become smaller than start date', () => {
        const value = '2005-01-01'
        endDateInput.simulate('change', { target: { value } })
        expect(
            wrapper.find(`span.uaa-date-input-error.${END_DATE}`)
        ).toHaveText(ERROR_END_BEFORE_START)
    })
    it('when endDate changes into range-error startDate error is cleared', () => {
        startDateInput.simulate('change', { target: { value: '2019-12-30' } })
        endDateInput.simulate('change', { target: { value: '2018-09-29' } })
        expect(
            wrapper.find(`span.uaa-date-input-error.${START_DATE}`).length
        ).toEqual(0)
    })
    it('when startDate changes into range-error the endDate error is cleared', () => {
        endDateInput.simulate('change', { target: { value: '2018-09-29' } })
        startDateInput.simulate('change', { target: { value: '2019-12-30' } })
        expect(
            wrapper.find(`span.uaa-date-input-error.${END_DATE}`).length
        ).toEqual(0)
    })
    it('when both fields get pattern errors on change both error fields are displayed', () => {
        startDateInput.simulate('change', { target: { value: NOT_A_DATE } })
        endDateInput.simulate('change', { target: { value: NOT_A_DATE } })
        expect(wrapper.find('span.uaa-date-input-error').length).toEqual(2)
    })
})
