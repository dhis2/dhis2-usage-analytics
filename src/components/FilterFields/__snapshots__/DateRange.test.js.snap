// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateRange/> matches the snapshot with default props 1`] = `
<DateRange
  endDate="2018-12-30"
  startDate="2018-09-30"
  updateFilter={[MockFunction]}
  updateUsageData={[MockFunction]}
>
  <div
    className="uaa-date-range"
  >
    <InputField
      className="startDate"
      dataTest="uaa-startdate"
      error={false}
      label="Start Date"
      name="startDate"
      onChange={[Function]}
      type="date"
      validationText={null}
      value="2018-09-30"
    >
      <Field
        className="startDate"
        dataTest="uaa-startdate"
      >
        <div
          className="jsx-1540540055 startDate"
          data-test="uaa-startdate"
        >
          <Label
            dataTest="dhis2-uicore-label"
            htmlFor="startDate"
          >
            <label
              className="jsx-3751407064 "
              data-test="dhis2-uicore-label"
              htmlFor="startDate"
            >
              <span
                className="jsx-3751407064"
              >
                Start Date
              </span>
              <JSXStyle
                id="3751407064"
              />
            </label>
          </Label>
          <Constrictor
            dataTest="dhis2-uicore-constrictor"
            minWidth="72px"
          >
            <div
              className="jsx-1696505344"
              data-test="dhis2-uicore-constrictor"
            >
              <Input
                dataTest="dhis2-uicore-input"
                error={false}
                name="startDate"
                onChange={[Function]}
                type="date"
                value="2018-09-30"
              >
                <div
                  className="jsx-3353877153 jsx-1090541828 input"
                  data-test="dhis2-uicore-input"
                >
                  <input
                    className="jsx-3353877153 jsx-1090541828 "
                    id="startDate"
                    name="startDate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="date"
                    value="2018-09-30"
                  />
                  <div
                    className="jsx-3353877153 jsx-1090541828 status-icon"
                  >
                    <StatusIcon
                      defaultTo={null}
                      error={false}
                    />
                  </div>
                  <JSXStyle
                    id="1090541828"
                  />
                  <JSXStyle
                    id="3353877153"
                  />
                </div>
              </Input>
              <JSXStyle
                dynamic={
                  Array [
                    "",
                    "min-width: 72px;",
                    "",
                  ]
                }
                id="1474598979"
              />
            </div>
          </Constrictor>
          <JSXStyle
            dynamic={
              Array [
                "16px",
              ]
            }
            id="618636905"
          />
        </div>
      </Field>
    </InputField>
    <InputField
      className="endDate"
      dataTest="uaa-enddate"
      error={false}
      label="End Date"
      name="endDate"
      onChange={[Function]}
      type="date"
      validationText={null}
      value="2018-12-30"
    >
      <Field
        className="endDate"
        dataTest="uaa-enddate"
      >
        <div
          className="jsx-1540540055 endDate"
          data-test="uaa-enddate"
        >
          <Label
            dataTest="dhis2-uicore-label"
            htmlFor="endDate"
          >
            <label
              className="jsx-3751407064 "
              data-test="dhis2-uicore-label"
              htmlFor="endDate"
            >
              <span
                className="jsx-3751407064"
              >
                End Date
              </span>
              <JSXStyle
                id="3751407064"
              />
            </label>
          </Label>
          <Constrictor
            dataTest="dhis2-uicore-constrictor"
            minWidth="72px"
          >
            <div
              className="jsx-1696505344"
              data-test="dhis2-uicore-constrictor"
            >
              <Input
                dataTest="dhis2-uicore-input"
                error={false}
                name="endDate"
                onChange={[Function]}
                type="date"
                value="2018-12-30"
              >
                <div
                  className="jsx-3353877153 jsx-1090541828 input"
                  data-test="dhis2-uicore-input"
                >
                  <input
                    className="jsx-3353877153 jsx-1090541828 "
                    id="endDate"
                    name="endDate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="date"
                    value="2018-12-30"
                  />
                  <div
                    className="jsx-3353877153 jsx-1090541828 status-icon"
                  >
                    <StatusIcon
                      defaultTo={null}
                      error={false}
                    />
                  </div>
                  <JSXStyle
                    id="1090541828"
                  />
                  <JSXStyle
                    id="3353877153"
                  />
                </div>
              </Input>
              <JSXStyle
                dynamic={
                  Array [
                    "",
                    "min-width: 72px;",
                    "",
                  ]
                }
                id="1474598979"
              />
            </div>
          </Constrictor>
          <JSXStyle
            dynamic={
              Array [
                "16px",
              ]
            }
            id="618636905"
          />
        </div>
      </Field>
    </InputField>
  </div>
</DateRange>
`;
